### ISSUE TITLE
Initial route not loading when running tests (2.4.0)

### ISSUE THREAD

**Issue by mark-apn** [Milad-Akarie/auto_route_library#710]
We tried upgrading to auto_route 2.4.0 yesterday but we ran into a bug regarding the updates around the route notifiers. 

Our app has 2 guards and all tests that deviate from the default initial route in a guard (e.g. not logged in) works fine. The only test that fails is the one where we want to load the correct initial page.

After digging around we noticed that the last print in the screenshot ('Notify route updated') is not called in that case.

This is located in routing_controller.dart in method `_pushAllGuarded`. It is going thru the else part with the `_updateSharedPathData` call, but somehow the test app is not inflating the route.

![image](https://user-images.githubusercontent.com/62654433/134129723-fe1583a9-03db-4d05-91ae-7e8efd2be980.png)

![image](https://user-images.githubusercontent.com/62654433/134129926-2ebdd3d8-498d-48d9-8125-4c14725c88ae.png)

**Comment by JeSuisAlrick** [comment]
Having this same issue.

**Comment by JeSuisAlrick** [comment]
@Milad-Akarie, are you looking into this?

**Comment by Milad-Akarie** [comment]
@mark-apn @JeSuisAlrick  I have no idea how to reproduce this, I'll have some time over the weekend, could you guys provide more information?

**Comment by zs-dima** [comment]
@Milad-Akarie I could see this issue as well

**Comment by Milad-Akarie** [comment]
@zs-dima in tests?

**Comment by zs-dima** [comment]
@Milad-Akarie in the applications, so I had to roll back to prev auto_route version.

**Comment by Milad-Akarie** [comment]
@zs-dima how to reproduce?

**Comment by zs-dima** [comment]
@Milad-Akarie projects quite big so I will be able to look into test project later

**Comment by Milad-Akarie** [comment]
@zs-dima Thanks.

**Comment by clement-buchart** [comment]
Hello @Milad-Akarie 
I was able to reproduce this behaviour with the mobile app from the example folder in this repo.

I simply added the `AuthGuard` to the HomePage in router.dart, and changed it's implementation to the following : 
```
class AuthGuard extends AutoRouteGuard {
  @override
  void onNavigation(NavigationResolver resolver, StackRouter router) {
    Future.delayed(Duration(seconds: 3)).then((value) => resolver.next(true));
  }
}
```

Then, the app stays indefinitely on a grey screen on boot.

It seems there is a race condition because in my real app, we are looking up a value in storage on boot to redirect to decide between redirecting to a tutorial or not. Sometimes everything works fine, but sometimes the app gets stuck on the root Router placeholder screen.

Hope this helps.

**Comment by JeSuisAlrick** [comment]
@clement-buchart, 

Good catch. 

In #703 I was wondering why AuthGuard was [removed from the root router](https://github.com/Milad-Akarie/auto_route_library/commit/c688c0c6095451984347380587ab05dc2ef03390#diff-242963ccd6808d464f8ae2907c24f830e4fd5f509f9e181ae1ffd2274698af20L20) and the [AutoRoute for BookDetailsPage](https://github.com/Milad-Akarie/auto_route_library/commit/c688c0c6095451984347380587ab05dc2ef03390#diff-feab261fb04545c9635174264cc85a5f8a45f516a3b6e210a4dfa5d127bb1f56L34). 

Isn't it suppose to be demonstrating the use of AuthGuard in the example?

**Comment by Milad-Akarie** [comment]
@JeSuisAlrick @clement-buchart  I think that was caused by the same route update issue,
should be fixed in  v2.4.1
please confirm and close.

**Comment by JeSuisAlrick** [comment]
> @JeSuisAlrick @clement-buchart I think that was caused by the same route update issue, should be fixed in v2.4.1 please confirm and close.

These changes work for me. Thanks @Milad-Akarie

**Comment by zs-dima** [comment]
@Milad-Akarie thanks a lot, works well now
